1.x: peer to peer replication with attachments does not work with android client

I tested peer to peer replication with attachment between net-mono and android device using the below command:
```
pytest -s -k test_peer_2_peer_sanity_push --liteserv-one-platform=net-mono 
--liteserv-one-version=1.4-50 --liteserv-one-host=<net_mono_ip> 
--liteserv-one-port=51000 --liteserv-one-storage-engine=SQLite 
--liteserv-two-platform=android --liteserv-two-version=1.4-44 
--liteserv-two-host=<Android_IP> --liteserv-two-port=52000 
--liteserv-two-storage-engine=SQLite  testsuites/listener/shared/client_client/
```

Without attachment, replication works fine. With attachment, replication does not work. I see the below trace in the logs:
```
02-27 15:46:47.456  2492  2534 W Router  : Router unable to route request to do_PUT_Document
02-27 15:46:47.456  2492  2534 W Router  : java.lang.reflect.InvocationTargetException
02-27 15:46:47.456  2492  2534 W Router  : 	at java.lang.reflect.Method.invoke(Native Method)
02-27 15:46:47.456  2492  2534 W Router  : 	at com.couchbase.lite.router.Router.start(Router.java:560)
02-27 15:46:47.456  2492  2534 W Router  : 	at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:147)
02-27 15:46:47.456  2492  2534 W Router  : 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
02-27 15:46:47.456  2492  2534 W Router  : 	at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2408)
02-27 15:46:47.456  2492  2534 W Router  : 	at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2330)
02-27 15:46:47.456  2492  2534 W Router  : 	at Acme.Serve.Serve$ServeConnection.run(Serve.java:2110)
02-27 15:46:47.456  2492  2534 W Router  : 	at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1238)
02-27 15:46:47.456  2492  2534 W Router  : 	at java.lang.Thread.run(Thread.java:761)
02-27 15:46:47.456  2492  2534 W Router  : Caused by: com.couchbase.lite.CouchbaseLiteException, Status: 406 (HTTP 406 not_acceptable)
02-27 15:46:47.456  2492  2534 W Router  : 	at com.couchbase.lite.router.Router.getBodyAsDictionary(Router.java:247)
02-27 15:46:47.456  2492  2534 W Router  : 	at com.couchbase.lite.router.Router.do_PUT_Document(Router.java:2364)
02-27 15:46:47.456  2492  2534 W Router  : 	... 9 more
```

It seems like android listener is not handling the multipart request.

--- a/src/main/java/com/couchbase/lite/router/Router.java
+++ b/src/main/java/com/couchbase/lite/router/Router.java
@@ -45,6 +45,7 @@ import com.couchbase.lite.replicator.Replication.ChangeListener;
 import com.couchbase.lite.replicator.Replication.ReplicationStatus;
 import com.couchbase.lite.replicator.ReplicationState;
 import com.couchbase.lite.storage.SQLException;
+import com.couchbase.lite.support.MultipartDocumentReader;
 import com.couchbase.lite.support.RevisionUtils;
 import com.couchbase.lite.support.Version;
 import com.couchbase.lite.util.Log;
@@ -86,7 +87,10 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
     private static final long MIN_CHANGES_HEARTBEAT = 5000;    // 5 seconds
     private static final long DEFAULT_CHANGES_TIMEOUT = 60000; // 60 seconds
 
+    private static final int BUF_LEN = 1024;
+
     private static final String CONTENT_TYPE_JSON = "application/json";
+    private static final String CONTENT_TYPE_MULTIPART = "multipart/";
 
     /**
      * Options for what metadata to include in document bodies
@@ -240,6 +244,32 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
         return eTag.equals(requestIfNoneMatch);
     }
 
+    // content-type: multipart/json
+    private Map<String, Object> getMultipartBodyAsDictionary() throws CouchbaseLiteException {
+        MultipartDocumentReader reader = new MultipartDocumentReader(db);
+        reader.setHeaders(Utils.headersToMap(connection.getRequestProperties()));
+        byte[] buffer = new byte[BUF_LEN];
+        int numBytesRead;
+        InputStream contentStream = connection.getRequestInputStream();
+        try {
+            while ((numBytesRead = contentStream.read(buffer)) != -1)
+                reader.appendData(buffer, 0, numBytesRead);
+        } catch (IOException ioe) {
+            Log.e(Log.TAG_ROUTER, "Error in load data from stream.");
+            throw new CouchbaseLiteException(Status.BAD_REQUEST);
+        } finally {
+            if (contentStream != null) {
+                try {
+                    contentStream.close();
+                } catch (IOException ioe) {
+                }
+            }
+        }
+        reader.finish();
+        return reader.getDocumentProperties();
+    }
+
+    // content-type: application/json
     private Map<String, Object> getBodyAsDictionary() throws CouchbaseLiteException {
         // check if content-type is `application/json`
         String contentType = getRequestHeaderContentType();
@@ -2363,12 +2393,18 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
 
     public Status do_PUT_Document(Database _db, String docID, String _attachmentName)
             throws CouchbaseLiteException {
-        Status status = new Status(Status.CREATED);
-        Map<String, Object> body = getBodyAsDictionary();
-        if (body == null) {
+        Map<String, Object> body;
+        String contentType = getRequestHeaderContentType();
+        // multipart
+        if (contentType != null && contentType.startsWith(CONTENT_TYPE_MULTIPART))
+            body = getMultipartBodyAsDictionary();
+            // application/json
+        else
+            body = getBodyAsDictionary();
+        if (body == null)
             throw new CouchbaseLiteException(Status.BAD_REQUEST);
-        }
 
+        Status status = new Status(Status.CREATED);
         if (getQuery("new_edits") == null ||
                 (getQuery("new_edits") != null && (Boolean.valueOf(getQuery("new_edits"))))) {
             // Regular PUT

--- a/src/main/java/com/couchbase/lite/util/Utils.java
+++ b/src/main/java/com/couchbase/lite/util/Utils.java
@@ -21,10 +21,13 @@ import com.couchbase.lite.replicator.RemoteRequestResponseException;
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
 import java.util.TreeMap;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
 import java.util.zip.GZIPInputStream;
 import java.util.zip.GZIPOutputStream;
 
@@ -278,6 +281,25 @@ public class Utils {
         Map<String, String> map = new TreeMap<String, String>(String.CASE_INSENSITIVE_ORDER);
         for (String name : headers.names()) {
             map.put(name, headers.get(name));
+        }
+        return map;
+    }
+
+    public static String join(List<String> list, String separator) {
+        StringBuilder sb = new StringBuilder();
+        for (String str : list) {
+            if (sb.length() != 0)
+                sb.append(separator);
+            sb.append(str);
+        }
+        return sb.toString();
+    }
+
+    public static Map<String, String> headersToMap(Map<String, List<String>> headers) {
+        List<String>a = new ArrayList<String>();
+        Map<String, String> map = new TreeMap<String, String>(String.CASE_INSENSITIVE_ORDER);
+        for (String name : headers.keySet()) {
+            map.put(name, join(headers.get(name), ", "));
         }
         return map;
     }

--- a/src/main/java/com/couchbase/lite/util/Utils.java
+++ b/src/main/java/com/couchbase/lite/util/Utils.java
@@ -296,7 +296,6 @@ public class Utils {
     }
 
     public static Map<String, String> headersToMap(Map<String, List<String>> headers) {
-        List<String>a = new ArrayList<String>();
         Map<String, String> map = new TreeMap<String, String>(String.CASE_INSENSITIVE_ORDER);
         for (String name : headers.keySet()) {
             map.put(name, join(headers.get(name), ", "));

